lib/formats/csv.flow

```diff
@@ -3,23 +3,125 @@ import ds/array;
+	str = flow2s(args[0]);
+	trimmed = substring(str, 1, strlen(str) - 2);
+	strReplace(trimmed, "\"\"", "\"");
+}));
+
+CsvRowInfo(start : int, end : int, row : [string]);
+
+parseCsvData(data : string, styles : [ParseCsvOption]) -> Pair<[[string]], string> {
+	styleSepM = tryExtractStruct(styles, ParseCsvSeparator(""));
+	sepStr = eitherFn(styleSepM, \styleSep -> styleSep.sep, \ -> {
+		if (containsStruct(styles, ParseCsvDetectSeparator())) {
+			sep = detectCsvSeparator(data).second;
+			devtrace("parseCsvData: detected separator is " + sep);
+			sep;
+		} else {
+			","; // default separator
+		}
+	});
+
+	// convert string to unicode escapes
+	sep = concatStrings(map(s2a(sepStr), \code -> {
+		hex = formatHex(code);
+		substring("\\u0000", 0, 6 - strlen(hex)) + hex;
+	}));
+
+	// cannot have double quotes or empty string as a separator
+	if (strContains(sep, "\\u0022") || sepStr == "") {
+		Pair([], "invalid separator");
+	} else {
+		grammarM = lookupTree(^compiledGrammarsR, sep);
+		grammar = eitherFn(grammarM, idfn, \ -> {
+			newOps = compilePegGrammar(formatString("#include formats/csv/csv.lingo-template", [sep]));
+			compiledGrammarsR := setTree(^compiledGrammarsR, sep, newOps);
+			newOps;
+		});
+
+		triple = parsic3(grammar, data, csvGrammarActions, []);
+		if (triple.third != "") {
+			devtrace("parseCsvData: " + triple.third);
+			Pair([], "parsing error");
+		} else {
+			array = map(triple.first, \info -> if (info.start == info.end) [] else info.row);
+			result = if (containsStruct(styles, ParseCsvSkipEmpty())) {
+				filter(array, \row -> row != []);
+			} else {
+				// trim empty rows at the end
+				idxM = lastfindi(array, \row -> row != []);
+				eitherMap(idxM, \idx -> take(array, idx + 1), array);
+			}
+
+			if (containsStruct(styles, ParseCsvRectangle())) {
+				numCols = uniq(map(result, length));
+				if (length(numCols) > 1) {
+					Pair(result, "inconsistent number of columns");
+				} else {
+					Pair(result, "");
+				}
+			} else {
+				Pair(result, "");
+			}
+		}
+	}
+}
+
+detectCsvSeparator(s : string) -> Pair<int, string> {
+	part = strLeft(s, 2000);
+	counts = map([",", ";", "\t"], \sep -> if (strCountOf(part, sep) > 0) getBestRowCount(part, sep) else 0);
+	if (counts[0] == counts[1] && counts[1] == counts[2]) Pair(counts[0], ",") else max3(
+		Pair(counts[0], ","),
+		Pair(counts[1], ";"),
+		Pair(counts[2], "\t"),
+	);
+}
+
+// How many rows with the given delimiter will have the same number of columns (the more, the more likely this is the delimiter)
+getBestRowCount(file : string, sep : string) -> int {
+	// Break the text into lines
+	lines = strSplit2WithoutLeave(file, [fromCharCode(10), fromCharCode(13)]);
+
+	// Calculate how many columns we'll have in each row
+	// Ignore rows with only one column in them, it means the delimiter is simply not there
+	arr_cnt = filter(map(map(lines, \a -> strSplitCheckQuotes(a, sep, [])), length), neq(1));
```

Comment:
too many array's iteration, you can do it in a single pass. Probably can be even grouped with the next line operations